### PR TITLE
Synchronize client detach and hook invocation

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Synchronize client detach and hook invocation

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: netebpfext/net_ebpf_ext.c [hunk 1]
```diff
@@ -632,6 +634,46 @@ _net_ebpf_ext_provider_client_attach_callback(
     return EBPF_SUCCESS;
 }
 
+static _Function_class_(KDEFERRED_ROUTINE) _IRQL_requires_max_(DISPATCH_LEVEL) _IRQL_requires_min_(DISPATCH_LEVEL)
+    _IRQL_requires_(DISPATCH_LEVEL) _IRQL_requires_same_ VOID _net_ebpf_ext_rundown(
+        _In_ struct _KDPC* Dpc,
+        _In_opt_ PVOID DeferredContext,
+        _In_opt_ PVOID SystemArgument1,
+        _In_opt_ PVOID SystemArgument2)
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:642]:
```suggestion
        _In_ struct _KDPC* dpc,
        _In_opt_ void* deferred_context,
        _In_opt_ void* system_argument1,
        _In_opt_ void* system_argument2)
```
use lower_snake_case for argument names just like for variables, and avoid P... types

File: netebpfext/net_ebpf_ext.c [hunk 2]
```diff
@@ -632,6 +634,46 @@ _net_ebpf_ext_provider_client_attach_callback(
...
+
+    UNREFERENCED_PARAMETER(Dpc);
+    UNREFERENCED_PARAMETER(SystemArgument1);
+    UNREFERENCED_PARAMETER(SystemArgument2);
+    if (registration)
+        KeSetEvent(&registration->rundown_wait, 0, FALSE);
+}
+
+static void
+_net_ebpf_ext_init_rundown(net_ebpf_ext_hook_provider_registration_t* registration)
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:655]: add SAL annotation (same for line 662, ...) ?

File: netebpfext/net_ebpf_ext.c [hunk 3]
```diff
@@ -632,6 +634,46 @@ _net_ebpf_ext_provider_client_attach_callback(
...
+}
+
+static void
+_net_ebpf_ext_wait_for_rundown(net_ebpf_ext_hook_provider_registration_t* registration)
+{
+    // Queue a DPC to each CPU and wait for it to run.
+    // After it has run on each CPU we can be sure that no
+    // DPC is busy processing a hook.
+    ULONG maximum_processor = KeQueryMaximumProcessorCount();
+    ULONG processor;
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:668]: can standard types like uint32_t be used here instead of ULONG or does that give a compiler error?

**Review by dthaler — APPROVED** [review]
Looks good to me now

File: netebpfext/net_ebpf_ext.c [hunk 4]
```diff
@@ -632,6 +634,46 @@ _net_ebpf_ext_provider_client_attach_callback(
...
+    for (processor = 0; processor < maximum_processor; processor++) {
+        KeSetTargetProcessorDpc(&registration->rundown_dpc, (uint8_t)processor);
+        if (KeInsertQueueDpc(&registration->rundown_dpc, NULL, NULL)) {
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext.c:671]: if the ebpf extension invokes the program on passive thread and it swapped out, and then this DPC ran, and the thread swapped back in, isnt that an issue?

Comment (Alan-Jowett) [reply netebpfext/net_ebpf_ext.c:671]: For hooks that run at passive, the extension would need to find an alternative approach.

**Pull request merged**